docs/plugin-development/index.md

```diff
@@ -676,6 +676,7 @@ Called at the end of `bundle.generate()` or immediately before the files are wri
 type AssetInfo = {
 	fileName: string;
 	name?: string;
+	needsCodeReference: boolean;
```

Comment:
Shouldn't this be optional for backwards compatibility? e.g. `needsCodeReference?` otherwise it will be a breaking change and will error on tsc build.

Reply:
Is this causing problems for you? `AssetInfo` is not a type that is written by the user but one returned by Rollup, I would expect ignoring the additional property should not cause issues.

Reply:
@lukastaegert unfortunately it is affecting me, as I write plugins and right now it is complaining about this missing prop `needsCodeReference` the above is just an example, sorry my point is anywhere that references `needsCodeReference` it needs to be made optional. Thank you.

Reply:
Could you show the wrong plugin code? Did you write a new type yourself?

Reply:
@TrickyPi

For example I have a function here:

```ts
import type {OutputBundle} from 'rollup';

export function addAsset(bundle: OutputBundle, fileName: string, source: string): void {
  bundle[fileName] = {type: 'asset', name: undefined, fileName, source};
}
```

Here is what tsc is complaining:
<img width="1066" alt="Screenshot 2023-01-31 at 8 35 29 pm" src="https://user-images.githubusercontent.com/1149825/215722961-8577724b-6054-4034-8bb7-f754e38c6d6f.png">

Reply:
That is something you should never do but maybe the documentation is not explicit enough: https://rollupjs.org/plugin-development/#generatebundle

> You can prevent files from being emitted by deleting them from the bundle object in this hook. To emit additional files, use the [this.emitFile](https://rollupjs.org/#this-emitfile) plugin context function

To state it differently: PLEASE always use `this.emitFile` to add files to the bundle, never add them directly. The reason is exactly your issue.

Reply:
@lukastaegert it was working before, but can you please explain what could go wrong if you add to the bundle directly? I've seen many big plugins do it as well: https://github.com/vitejs/vite/blob/main/packages/plugin-legacy/src/index.ts#L708

Reply:
Yes, and they should not do it. Unfortunately, I did not put big enough exclamation marks into the docs in that section. If you manually add something to the bundle, you are circumventing any internal mechanisms that Rollup has of tracking assets, which can cause any number of issues if Rollup uses any caching here. Also, your plugin will not respect any changes in the properties that Rollup adds to assets, as you just noticed. There is only one officially supported way of adding assets or any kind of file and that is through `this.emitFile`. For any other way, be prepared for breaking changes in minor versions.

Reply:
Here you go: #4831